apm-agent-core/src/test/java/co/elastic/apm/agent/configuration/CoreConfigurationTest.java

```diff
@@ -0,0 +1,120 @@
+            assertThat(config.isInstrumentationEnabled(Arrays.asList("foo", "bar"))).isFalse();
+        } finally {
+            systemProperties.remove("enable_instrumentations");
+            systemProperties.remove("disable_instrumentations");
+        }
+    }
+
+    @Test
+    void testWithEnabledInstrumentations() {
+        systemProperties.put("enable_instrumentations", "foo");
+        systemProperties.put("disable_instrumentations", "");
+        try {
+            CoreConfiguration config = getCoreConfiguration();
+            assertThat(config.isInstrumentationEnabled("foo")).isTrue();
+            assertThat(config.isInstrumentationEnabled("bar")).isFalse();
+            assertThat(config.isInstrumentationEnabled(Collections.singletonList("foo"))).isTrue();
+            assertThat(config.isInstrumentationEnabled(Collections.singletonList("bar"))).isFalse();
+            assertThat(config.isInstrumentationEnabled(Arrays.asList("foo", "bar"))).isTrue();
+        } finally {
+            systemProperties.remove("enable_instrumentations");
+            systemProperties.remove("disable_instrumentations");
+        }
+    }
+
+    @Test
+    void testWithDisabledAndEnabledInstrumentations() {
+        systemProperties.put("enable_instrumentations", "foo");
+        systemProperties.put("disable_instrumentations", "foo");
+        try {
+            CoreConfiguration config = getCoreConfiguration();
+            assertThat(config.isInstrumentationEnabled("foo")).isFalse();
+            assertThat(config.isInstrumentationEnabled("bar")).isFalse();
+            assertThat(config.isInstrumentationEnabled(Collections.singletonList("foo"))).isFalse();
+            assertThat(config.isInstrumentationEnabled(Collections.singletonList("bar"))).isFalse();
+            assertThat(config.isInstrumentationEnabled(Arrays.asList("foo", "bar"))).isFalse();
+        } finally {
+            systemProperties.remove("enable_instrumentations");
+            systemProperties.remove("disable_instrumentations");
+        }
+    }
+
+    @Test
+    void testLegacyDefaultDisabledInstrumentation() {
+        systemProperties.put("enable_instrumentations", "");
+        systemProperties.put("disable_instrumentations", "incubating");
+        try {
+            CoreConfiguration config = getCoreConfiguration();
+            assertThat(config.isInstrumentationEnabled("experimental")).isFalse();
+            assertThat(config.isInstrumentationEnabled(Collections.singletonList("experimental"))).isFalse();
+        } finally {
+            systemProperties.remove("enable_instrumentations");
+            systemProperties.remove("disable_instrumentations");
+        }
+    }
+
+    private static CoreConfiguration getCoreConfiguration() {
+        ConfigurationRegistry configurationRegistry = ConfigurationRegistry.builder().addOptionProvider(new CoreConfiguration()).addConfigSource(new SystemPropertyConfigurationSource()).build();
```

Comment:
Instead of "polluting" the system properties which you have to clear in the finally block add a `SimpleSource` parameter to this method. The different tests can then call `getCoreConfiguration(SimpleSource.of("enable_instrumentations", "").and("disable_instrumentations", "incubating")`, for example.

Reply:
Thank You! That makes the code much nicer!